File: android_native/app/src/main/java/com/ezbillify/ezworkspace/services/AuthService.kt
package com.ezbillify.ezworkspace.services

import com.ezbillify.ezworkspace.EzWorkspaceApp
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.exceptions.RestException
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull

/** Employee profile loaded from the shared `employees` table. */
data class AppUser(
    val id: String,
    val name: String,
    val email: String,
    val role: String = "employee",
    val designation: String? = null,
    val department: String? = null,
) {
    val initials: String
        get() {
            val parts = name.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
            if (parts.isEmpty()) return "?"
            if (parts.size == 1) return parts[0].take(1).uppercase()
            return (parts.first().take(1) + parts.last().take(1)).uppercase()
        }
}

class AuthService {
    private val client get() = EzWorkspaceApp.supabase

    val isLoggedIn: Boolean
        get() = client.auth.currentSessionOrNull() != null

    /** Signs in against the SAME Supabase project the web app and Flutter build use. */
    suspend fun signIn(email: String, password: String): AppUser {
        val clean = email.trim().lowercase()
        client.auth.signInWith(Email) {
            this.email = clean
            this.password = password
        }
        return loadProfile() ?: AppUser(
            id = client.auth.currentUserOrNull()?.id ?: "",
            name = nameFromEmail(clean),
            email = clean,
        )
    }

    /** Loads the employee profile for the current auth user (RLS-permitting). */
    suspend fun loadProfile(): AppUser? {
        val user = client.auth.currentUserOrNull() ?: return null
        val email = (user.email ?: "").lowercase()
        return try {
            val row = client.from("employees")
                .select(Columns.raw("id, name, email, role, designation, department")) {
                    filter {
                        or {
                            ilike("email", email)
                            ilike("personal_email", email)
                            ilike("zoho_email", email)
                        }
                    }
                }
                .decodeSingleOrNull<JsonObject>() ?: return null

            fun str(key: String) = row[key]?.jsonPrimitive?.contentOrNull

            AppUser(
                id = str("id") ?: user.id,
                name = str("name") ?: nameFromEmail(email),
                email = str("email") ?: email,
                role = str("role") ?: "employee",
                designation = str("designation"),
                department = str("department"),
            )
        } catch (_: RestException) {
            // RLS may block the read — fall back to auth identity so login still works.
            AppUser(id = user.id, name = nameFromEmail(email), email = email)
        } catch (_: Exception) {
            AppUser(id = user.id, name = nameFromEmail(email), email = email)
        }
    }

    suspend fun signOut() {
        client.auth.signOut()
    }

    private fun nameFromEmail(email: String): String {
        val local = email.substringBefore("@").replace(Regex("[._]"), " ")
        return local.split(" ")
            .filter { it.isNotEmpty() }
            .joinToString(" ") { it.replaceFirstChar(Char::uppercase) }
    }
}

File: flutter_app/lib/services/auth_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

/// Employee profile loaded from the shared `employees` table.
class AppUser {
  AppUser({
    required this.id,
    required this.name,
    required this.email,
    this.role = 'employee',
    this.designation,
    this.department,
  });

  final String id;
  final String name;
  final String email;
  final String role;
  final String? designation;
  final String? department;

  String get initials {
    final parts = name.trim().split(RegExp(r'\s+'));
    if (parts.isEmpty || parts.first.isEmpty) return '?';
    if (parts.length == 1) return parts.first.substring(0, 1).toUpperCase();
    return (parts.first[0] + parts.last[0]).toUpperCase();
  }
}

class AuthService {
  final SupabaseClient _sb = Supabase.instance.client;

  Session? get session => _sb.auth.currentSession;
  bool get isLoggedIn => session != null;

  /// Signs in against the SAME Supabase project the web app uses.
  Future<AppUser> signIn(String email, String password) async {
    final clean = email.trim().toLowerCase();
    final res = await _sb.auth.signInWithPassword(email: clean, password: password);
    if (res.user == null) {
      throw const AuthException('Invalid email or password.');
    }
    return await loadProfile() ??
        AppUser(id: res.user!.id, name: _nameFromEmail(clean), email: clean);
  }

  /// Loads the employee profile for the current auth user (RLS-permitting).
  Future<AppUser?> loadProfile() async {
    final u = _sb.auth.currentUser;
    if (u == null) return null;
    final email = (u.email ?? '').toLowerCase();
    try {
      final row = await _sb
          .from('employees')
          .select('id, name, email, role, designation, department')
          .or('email.ilike.$email,personal_email.ilike.$email,zoho_email.ilike.$email')
          .maybeSingle();
      if (row == null) return null;
      return AppUser(
        id: row['id'].toString(),
        name: (row['name'] ?? _nameFromEmail(email)).toString(),
        email: (row['email'] ?? email).toString(),
        role: (row['role'] ?? 'employee').toString(),
        designation: row['designation']?.toString(),
        department: row['department']?.toString(),
      );
    } catch (_) {
      // RLS may block the read — fall back to auth identity so login still works.
      return AppUser(id: u.id, name: _nameFromEmail(email), email: email);
    }
  }

  Future<void> signOut() => _sb.auth.signOut();

  String _nameFromEmail(String email) {
    final local = email.split('@').first.replaceAll(RegExp(r'[._]'), ' ');
    return local
        .split(' ')
        .where((w) => w.isNotEmpty)
        .map((w) => w[0].toUpperCase() + w.substring(1))
        .join(' ');
  }
}
